### PR TITLE
Add Balancer weighted pool price oracle

### PR DESCRIPTION
**Pull request by fedgiac**
Adds a new price oracle for Balancer weighted pools.

Some notes:
- It should support all weighted pools from v1 to v4 since they all support the interface used in this PR.
- Unfortunately, it also returns a price for other pool types apart from weighted pools (e.g., managed pools). Determining from an id its pool type is difficult and would require checking the contract code for the old weighted pool versions and on-chain JSON manipulation for newer pool versions. To avoid the complexity (and corresponding gas cost) this isn't done and we trust the user to verify that the pool is a Balancer weighted pool.  

### How to test

New unit tests.

### CODE REVIEW

**Review by mfw78 — APPROVED** [review]
Minor nits. Verified that the logic and math from the perspective of being indicative to _signal_ orders (in the context of CoW AMM), however exhaustive derivation of formulae has not been taken into account and this should **NOT** be used as a price oracle where funds are subsequently at risk (fund security is done separately within CoW AMM).

File: src/interfaces/IBalancer.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     * be made for the difference between the requested amount and Internal Balance (if any). Note that ETH cannot be
+     * withdrawn from Internal Balance: attempting to do so will trigger a revert.
+     *
+     * This causes the Vault to call the `IBasePool.onJoinPool` hook on the Pool's contract, where Pools implement
+     * their own custom logic. This typically requires additional information from the user (such as the expected number
+     * of Pool shares). This can be encoded in the `userData` argument, which is ignored by the Vault and passed
+     * directly to the Pool's contract, as is `recipient`.
+     *
+     * Emits a `PoolBalanceChanged` event.
+     */
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/interfaces/IBalancer.sol]: Thanks for spotting this!
b7e98ce77a14da0221feca0e21df5d1a04169aae

File: src/oracles/BalancerWeightedPoolPriceOracle.sol [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+     * If truncating meant that less significan bits than `TOLERANCE` remained,
+     * then this function truncates less to preserve `TOLERANCE` bits in the
+     * smallest value, even if one of the output values ends up having more than
+     * 128 bits of size.
+     * @param num1 First input value.
+     * @param num2 Second input value.
+     * @return The two original input values in the original order with some of
+     * the least significant bits truncated.
+     */
+    function reduceOutputBytes(uint256 num1, uint256 num2) internal pure returns (uint256, uint256) {
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/oracles/BalancerWeightedPoolPriceOracle.sol:132]: 42ffc8291b0aa16f9f35844012c2da36f1f235b9

File: src/oracles/BalancerWeightedPoolPriceOracle.sol [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        uint256 priceDenominator = balanceToken1 * weightToken0;
+
+        return reduceOutputBytes(priceNumerator, priceDenominator);
+    }
+
+    /**
+     * @dev The two input values are truncated off their least significant bits
+     * by the same number of bits while trying to make them fit 128 bits.
+     * The number of bits that is truncated is always the same for both values.
+     * If truncating meant that less significan bits than `TOLERANCE` remained,
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/oracles/BalancerWeightedPoolPriceOracle.sol]: 96b16069ee2320c861af912e618b542c08457427

**Review by mfw78 — APPROVED** [review]

**Pull request merged**